### PR TITLE
Add acceleration structure building validation to GPU validation

### PR DESCRIPTION
**Pull request by jmacnak-nv**
Adds validation to detect invalid bottom level acceleration structure handles being used when building top level acceleration structures. The validation is a little more complicated because the instance buffer for vkCmdBuildAccelerationStructureNV() is potentially residing on the GPU. This is also why these changes put the validation in the GPU validation section.

I still need to update the GPU validation doc.

### CODE REVIEW

**Comment by mark-lunarg** [comment]
@tony, this passed CI, at least before your changes.  See change 4974.

File: layers/gpu_validation.cpp [hunk 1]
```diff
@@ -356,11 +597,730 @@ void CoreChecks::GpuPreCallRecordDestroyDevice() {
...
     gpu_validation_state->desc_set_manager.reset();
+
+    GpuDestroyAccelerationStructureBuildValidationState();
+
     if (gpu_validation_state->vmaAllocator) {
         vmaDestroyAllocator(gpu_validation_state->vmaAllocator);
     }
 }
 
+bool GpuAccelerationStructureBuildValidationState::GetAvailableDescriptorSet(
```
Comment (Tony-LunarG) [inline layers/gpu_validation.cpp]: Is the existing GpuDescriptorSetManager not sufficient for this?

File: layers/gpu_validation.cpp [hunk 2]
```diff
@@ -356,11 +597,730 @@ void CoreChecks::GpuPreCallRecordDestroyDevice() {
...
+
+        result = DispatchBeginCommandBuffer(command_buffer, &command_buffer_bi);
+        if (result != VK_SUCCESS) {
+            ReportSetupProblem(VK_DEBUG_REPORT_OBJECT_TYPE_DEVICE_EXT, HandleToUint64(device),
+                               "Failed to begin command buffer for acceleration structure build validation.");
+        }
+    }
+
+    if (result == VK_SUCCESS) {
+        DispatchCmdBuildAccelerationStructureNV(command_buffer, &as_ci.info, VK_NULL_HANDLE, 0, VK_FALSE,
```
Comment (Tony-LunarG) [inline layers/gpu_validation.cpp]: This call gets a validation error from core_validation line 6926
`if (scratch_buffer_state != nullptr && dst_as_state != nullptr &&
            dst_as_state->build_scratch_memory_requirements.memoryRequirements.size >
                (scratch_buffer_state->binding.size - (scratch_buffer_state->binding.offset + scratchOffset)))`
and I think the validation code is wrong.  I don't think scratch_buffer_state->binding.offset needs to be there.  The spec says just that the size must be less than or equal to the size of scratch minus scratchOffset.

File: tests/vklayertests_others.cpp [hunk 3]
```diff
@@ -4628,3 +4655,539 @@ TEST_F(VkLayerTest, ValidateCmdCopyAccelerationStructureNV) {
...
+    submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+    submit_info.commandBufferCount = 1;
+    submit_info.pCommandBuffers = &command_buffer.handle();
+    vkQueueSubmit(m_device->m_queue, 1, &submit_info, VK_NULL_HANDLE);
+    vkQueueWaitIdle(m_device->m_queue);
+    m_errorMonitor->VerifyFound();
+}
+
+TEST_F(VkLayerTest, GpuBuildAccelerationStructureValidationRestoresState) {
+    TEST_DESCRIPTION("Validate that acceleration structure gpu validation correclty restores compute state.");
```
Comment (Tony-LunarG) [inline tests/vklayertests_others.cpp]: Spelling - "correctly"

File: layers/gpu_validation.cpp (hunk 1 above)
Comment (jmacnak-nv) [reply layers/gpu_validation.cpp]: Updated to use GpuDescriptorSetManager. PTAL.

File: layers/gpu_validation.cpp (hunk 2 above)
Comment (jmacnak-nv) [reply layers/gpu_validation.cpp]: Removed. I mistakenly thought that scratch_buffer_state->binding.size was the size of the underlying VkDeviceMemory allocation and that the offset within the underlying memory needed to be calculated.

File: tests/vklayertests_others.cpp (hunk 3 above)
Comment (jmacnak-nv) [reply tests/vklayertests_others.cpp]: Fixed.

File: layers/gpu_validation.cpp (hunk 1 above)
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp]: Looks good

**Comment by Tony-LunarG** [comment]
LGTM  @jmacnak-nv were you going to update the documentation as part of this PR or in a future PR?

**Comment by jmacnak-nv** [comment]
Thanks for the review! I will try to get to that later today and update as part of this PR.

**Comment by jmacnak-nv** [comment]
Sorry for the delay.

Please take a look at the latest commit. I tried to move most of the generic layer information to the beginning of the document with feature specific documentation following. I then moved most of the existing documentation about the descriptor indexing shader checks into a feature sub section and added an additional sub section for the new acceleration structure building check.

File: docs/gpu_validation.md [hunk 4]
```diff
@@ -670,7 +686,7 @@ The Stage is the integer value used in SPIR-V for each of the Execution Models:
 |MissNV         |5317   |
 |CallableNV     |5318   |
 
-### Stage Specific Words
+####### Stage Specific Words
```
Comment (Tony-LunarG) [inline docs/gpu_validation.md]: Seven pound signs, at least on my markdown previewer, do not work as a header indication and instead appear as #######

Comment (Tony-LunarG) [reply docs/gpu_validation.md]: The same as true for the github previewer, so I think it needs to be addressed

**Comment by KarenGhavam-lunarG** [comment]
I will add that an SDK is going to be released next week and getting this resolved by Friday of this week will enable it's inclusion in the SDK.

**Comment by Tony-LunarG** [comment]
Rebased in #1242

**Comment by Tony-LunarG** [comment]
Merged via #1242

**Pull request closed without merging**